src/search_file.c

```diff
@@ -276,25 +279,13 @@ static bool file_matches(const char *filename, const char *search_term)
 	return regexec(&regexp_comp, filename, 0, NULL, 0) == 0;
 }
 
-static int search_in_manifest(struct manifest *mom, struct file *manifest_file, const char *search_term)
+static int search_in_manifest(struct manifest *mom, struct manifest *manifest, const char *search_term)
 {
 	struct list *l;
 	int count = 0;
-	struct manifest *m;
 	struct list *files = NULL;
 
-	// If manifest was not downloaded in download_all_manifests, skip this
-	// to avoid trying to download it again
-	if (!list_search(manifest_header_cache, manifest_file->filename, manifest_str_cmp)) {
-		return -ENOENT;
-	}
-
-	m = load_manifest(manifest_file->last_change, manifest_file, mom, false, NULL);
-	if (!m) {
-		return -ENOENT;
-	}
-
-	for (l = m->files; l; l = l->next) {
+	for (l = manifest->files; l; l = l->next) {
```

Comment:
we've had similar discussions before, but wouldn't it be a good idea to ue list_head to make sure we look at the head of the list? or should we trust the caller?

Reply:
We need to define who is in charge on keeping the list pointer to head. There are millions of situations where we don't call list_head() and I personally don't think it's a good idea to call it every time just in case. 

My suggestion is:
 - All functions from list.h should work if you pass a pointer to any position of the list. (so we should always call list_head).
 - All functions from list that is suppose to return a list should return the first element of the list. If a function return one item of the list we can return that item (list_append_data, list_remote_item, for example)
 - All functions that receives a list as parameter should expect to receive the first element of the list. 
 - All functions that returns a list should return the first element of the list.

By doing that we are always going to be pointing to the first element of the list. We just need the list_head in the first place because of functions that were returning a list pointing to random positions of the list.

By doing that we are going to be closer (but not there yet) to more used implementations of double linked lists. I would still prefer that functions from list API didn't call list_head, but this would be a big change and not much gain.

If you all agree I can attack this in a new PR.

Reply:
I like the idea, let's do it